src/client.rs

```diff
@@ -0,0 +1,94 @@
+use crate::config::Config;
+
+use reqwest::blocking::RequestBuilder;
+use reqwest::Result;
+use reqwest::header;
+use serde::Deserialize;
+use std::process;
+
+const STATUS_URL: &str = "https://gw.lamarzocco.io/v1/home/machines/LM035401/status";
+const AUTH_URL: &str = "https://cms.lamarzocco.io/oauth/v2/token";
+
+#[derive(Deserialize)]
+struct StatusResponse {
+    data: Machine
+}
+
+#[derive(Deserialize)]
+struct Machine {
+    #[serde(rename = "MACHINE_STATUS")]
+    machine_status: String
+}
+
+pub fn get_status(cfg: Config) -> Result<String> {
+    let client = reqwest::blocking::Client::new();
+
+    println!("Fetching status...");
+    let resp = put_headers(cfg, client.get(STATUS_URL))
+        .send()?
+        .json::<StatusResponse>();
+
+    match resp {
+        Ok(status) => return Ok(status.data.machine_status),
+        Err(err) => return Err(err),
+    };
+}
+
+pub fn put_status(cfg: Config, on: bool) -> Result<String> {
+    let status = if on { "ON" } else { "STANDBY" };
+    let body = String::from(format!("{{\"status\":\"{}\"}}", status));
+
+    println!("Updating status...");
+    let client = reqwest::blocking::Client::new();
+    let result = put_headers(cfg, client.get(STATUS_URL))
+        .json(&body)
+        .send()?
+        .text();
+
+    match result {
+        Ok(result) => return Ok(result),
+        Err(err) => return Err(err),
+    };
+}
+
+#[derive(Deserialize)]
+struct AccessToken {
+    access_token: String,
+}
+
+pub fn get_auth_token(cfg: Config) -> Result<String> {
+    let params = [
+        ("client_id", cfg.client_id),
+        ("client_secret", cfg.client_secret),
+        ("grant_type", String::from("password")),
+        ("password", cfg.password),
+        ("username", cfg.username),
+    ];
```

Comment:
Just for kicks you can do a [`impl Into<BTreeMap<String, String>> for Config`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html) or `Into<HashMap<String, String>>`(depending of if you care about order) allowing you to do a bit of tidying:
```suggestion
    let mut params: BTreeMap<_, _> = cfg.into()
    params.insert("grant_type".to_string(), "password".to_string());
```

Reply:
@mkatychev Elements from `Config` that I wouldn't want to send to the API would be included in the params 👎

Reply:
🤷‍♂️